Add the "internal" error type

toji opened the pull request:
~Fixes~Issue: #2308 (EDIT(@kainino0x): still need to keep that issue open about exposing non-uncategorized errors)

This error type captures non-validation errors that occur at a system
or implementation level. (Such as shaders failing because they are
too complex.) Initially only pipeline creation may produce operation
errors.

kainino0x commented on spec/index.bs:
```diff
@@ -6313,6 +6314,12 @@ dictionary GPUComputePipelineDescriptor : GPUPipelineDescriptorBase {
+                        [=shader-creation error|shader creation=] of
+                        |descriptor|.{{GPUComputePipelineDescriptor/compute}}.{{GPUProgrammableStage/module}}, or during
+                        the [=pipeline-creation error|pipeline creation=] of |pipeline|:
+                        1. [$generate an operation error$] and make |pipeline| [=invalid=].
```
There should also be something in createShaderModule, I think along the lines of "if there's an uncategorized shader error, make the GPUShaderModule invalid without generating any error"

kainino0x reviewed (commented):
I think this makes sense, modulo any possible changes to the error scope API which are pretty much orthogonal.

toji commented:
After editors discussion on Monday, we determined that there wasn't a solid need for a separate "operation" and "out-of-memory" error type, as the situations in which they could arise were operation specific and didn't overlap. As a result, it's cleaner to simply extend the existing "out-of-memory" errors to encompass the types of uncategorized errors described in the WGSL spec and rename that error type to "internal" to reflect it's new scope. This PR has been updated accordingly.

In the future should we see a need to differentiate OOM errors from other types of internal errors explicitly, we can add enums, flags, or other metadata to the `GPUInternalError` interface. It's not clear at this point, however, if that's actually beneficial.

kainino0x commented:
Notably it didn't seem to make sense to have a generic "internal" error as well as an "out-of-memory" error. It would make more sense to have "pipeline-failure" and "out-of-memory" or just "internal". We went with just "internal" for now because it's easy to expose the difference later as a field of GPUInternalError.

kainino0x commented on spec/index.bs:
```diff
@@ -12090,19 +12090,63 @@ of WebGPU calls, typically for debugging purposes or to make an operation more f
 };
 </script>
 
+{{GPUErrorFilter}} defines the type of errors that should be caught when calilng
```
```suggestion
{{GPUErrorFilter}} defines the type of errors that should be caught when calling
```

kainino0x commented on spec/index.bs:
```diff
@@ -12090,19 +12090,63 @@ of WebGPU calls, typically for debugging purposes or to make an operation more f
+    : <dfn>"validation"</dfn>
+    ::
+        Indicates that an operation did not satisfy all validation requirements. Validation errors
+        are always indicitive of an application error, and is expected to fail the same way across
```
```suggestion
        are always indicative of an application error, and is expected to fail the same way across
```

kdashg commented:
I think it's useful to preserve out-of-memory whenever we can tell that that's a reasonable possibility.

kainino0x commented on spec/index.bs:
```diff
@@ -12090,19 +12090,63 @@ of WebGPU calls, typically for debugging purposes or to make an operation more f
+or buffer, can be used to detect failures such as out of memory conditions, in which case the
+application may try freeing some resources and trying the allocation again.
+
+Error scopes do not indetify which command failed, however. So, for instance, wrapping all the
```
```suggestion
Error scopes do not identify which command failed, however. So, for instance, wrapping all the
```

kainino0x commented on spec/index.bs:
```diff
@@ -12090,19 +12090,63 @@ of WebGPU calls, typically for debugging purposes or to make an operation more f
+determine if the issue was due to memory constraints. As a result freeing resources would usually
+not be a productive response to a failure of that scope. A more appropriate response would be to
+allow the application to fall back to a different model or produce a warning that the model could
+not be loaded.
```
Perhaps suggest here that a smaller nested error scope could be used for this.

toji replied:
Done!

kainino0x commented:
> I think it's useful to preserve out-of-memory whenever we can tell that that's a reasonable possibility.

@kdashg out of memory errors can only come from createBuffer, createTexture, and createQuerySet. If you capture an internal error from one of those calls then you know it's an out of memory condition. Do you think that doesn't work well?

toji commented:
Updated to add a `GPUInternalErrorReason` enum, of which the only value is currently `"out-of-memory"` to satisfy Kelsey's request today.

Also, adding that created _just enough_ clutter in that section that I felt the urge to restructure it for clarity. So now there's one subsection that just describes the different `GPUError` variants, their purpose, and their members, and another subsection that just describes error scopes. I think it's a lot easier to follow now, but it does make the diff a bit messy. Sorry!

kainino0x reviewed (approved):
LGTM % nit

kainino0x commented on spec/index.bs:
```diff
@@ -12071,77 +12071,131 @@ partial interface GPUDevice {
+system or implementation-specific reason even when all validation requirements have been satisfied.
+For example, the operation may exceede the capabilities of the implementation in a way not easily
+captured by the [=supported limits=]. The same operation may succeed on other devices or under
+difference circumstances.
```
We should say that internal errors can only come from certain operations described in the spec, not any arbitrary operation.

toji replied:
Done, and also added an algorithm for generating an internal error for future use (and while I was at it cleaned up some of the algorithms to use the new timeline styles.)

Haven't actually added any instances of an operation generating an internal error yet, because this CL is already messy enough as is and I'd like those to be more easily reviewed.

Out of curiosity, what's the expected behavior for an operation that we don't explicitly say can generate internal error experiencing an internal error? Is that just forced to be a lost context every time?

kainino0x replied:
I didn't realize we didn't currently specify that out of memory can come from createBuffer/Texture/QuerySet. That'll be nice to have written down.

Yeah, the behavior I have in mind is:
- Implementation does any heroics it wants to try to do to make the error go away (like waiting for the queue to finish and then deleting any temporary resources) and tries again
- If it still fails, lose the device

toji commented on spec/index.bs:
```diff
@@ -12071,122 +12071,131 @@ partial interface GPUDevice {
 Note: {{GPUError}} may gain new subtypes in future versions of this spec. Applications should handle
 this possibility, using only the error's {{GPUError/message}} when possible, and specializing using
 `instanceof`. Use `error.constructor.name` when it's necessary to serialize an error (e.g. into
 JSON, for a debug report).
```
Done.

kainino0x commented on spec/index.bs:
```diff
@@ -12131,17 +12099,132 @@ JSON, for a debug report).
+
+<div algorithm>
+    To <dfn abstract-op lt="Generate an internal error|generate an internal error">generate an
+    internal error</dfn> with {{GPUInternalErrorReason}} |reason| for {{GPUDevice}} |device|, run
```
```suggestion
    internal error</dfn> with ({{GPUInternalErrorReason}} or [=undefined=]) |reason| for {{GPUDevice}} |device|, run
```

kainino0x commented on spec/index.bs:
```diff
@@ -12131,17 +12099,132 @@ JSON, for a debug report).
+
+<div algorithm>
+    To <dfn abstract-op lt="Generate an internal error|generate an internal error">generate an
+    internal error</dfn> with ({{GPUInternalErrorReason}} or [=undefined=]) |reason| for {{GPUDevice}} |device|, run
```
my bad this xlink wasn't correct
```suggestion
    internal error</dfn> with ({{GPUInternalErrorReason}} or undefined) |reason| for {{GPUDevice}} |device|, run
```

toji commented:
Editors meeting: After discussion about usage and addressing concerns raised by @kdashg and others, decided to break up "internal" and "out-of-memory" errors into separate types once again.

One motivating example for this change was the idea of developers who may want to capture an entire frame in an OOM error scope and use it as a signal to take some general action to reduce memory use, such as flushing caches. Combining internal and OOM errors would make that more difficult.

toji commented:
This approach was agreed on in today's meeting with the understanding that it satisfies @kdashg's previously voiced requirement (easily differentiating OOM errors). Merging.
